cdap-spark-core-base/src/main/java/io/cdap/cdap/app/runtime/spark/SparkPackageUtils.java

```diff
@@ -462,32 +465,69 @@ private static SparkFramework prepareSparkFramework(SparkCompat sparkCompat,
     String archiveName = "spark.archive-" + sparkVersion + "-" + VersionInfo.getVersion() + ".zip";
     Location frameworkDir = locationFactory.create("/framework/spark");
     Location frameworkLocation = frameworkDir.append(archiveName);
-
-    if (!frameworkLocation.exists()) {
-      File archive = new File(tempDir, archiveName);
-      try {
-        try (ZipOutputStream zipOutput = new ZipOutputStream(new BufferedOutputStream(new FileOutputStream(archive)))) {
-          zipOutput.setLevel(Deflater.NO_COMPRESSION);
-          for (File file : getLocalSparkLibrary(sparkCompat)) {
-            zipOutput.putNextEntry(new ZipEntry(file.getName()));
-            Files.copy(file.toPath(), zipOutput);
-            zipOutput.closeEntry();
-          }
+    File archive = new File(tempDir, archiveName);
+    try {
+      try (
+        ZipOutputStream zipOutput = new ZipOutputStream(new BufferedOutputStream(new FileOutputStream(archive)))) {
+        zipOutput.setLevel(Deflater.NO_COMPRESSION);
+        for (File file : getLocalSparkLibrary(sparkCompat)) {
+          zipOutput.putNextEntry(new ZipEntry(file.getName()));
+          Files.copy(file.toPath(), zipOutput);
+          zipOutput.closeEntry();
         }
+      } 
+      Location writeLockLocation = frameworkDir.append("write_in_progress");
+      frameworkDir.mkdirs("755");
 
-        // Upload spark archive to the framework location
-        frameworkDir.mkdirs("755");
-
-        try (OutputStream os = frameworkLocation.getOutputStream("644")) {
-          Files.copy(archive.toPath(), os);
+      while (!frameworkLocation.exists()) {
+        if (!uploadToLocation(frameworkLocation, archive, writeLockLocation)) {
+          waitForLocation(frameworkLocation, writeLockLocation);
         }
-      } finally {
-        archive.delete();
       }
+    } finally {
+      Files.deleteIfExists(archive.toPath());
     }
+
     return new SparkFramework(new LocalizeResource(resolveURI(frameworkLocation), true), SPARK_YARN_ARCHIVE);
   }
 
+  private static boolean uploadToLocation(Location frameworkLocation, File archive, Location writeLockLocation)
+    throws IOException {
+    if (!writeLockLocation.createNew()) {
+      return false;
+    }
+
+    Location tempArchiveHdfs = frameworkLocation.getTempFile("temp_jar");
+    try {
+      //Copy file to temp loc ( Upload )
+      try (OutputStream os = tempArchiveHdfs.getOutputStream()) {
+        Files.copy(archive.toPath(), os);
+      }
+      //Move to actual location
+      tempArchiveHdfs.renameTo(frameworkLocation);
+      writeLockLocation.delete();
+    } finally {
+      tempArchiveHdfs.delete();
+      archive.delete();
```

Comment:
Shouldn't have this as it is already covered by the finally block in the caller. It's better to have create and delete logic in the same method.